provider/source/src/time_zones/mod.rs

```diff
@@ -5,106 +5,143 @@
 use crate::cldr_serde;
 use crate::IterableDataProviderCached;
 use crate::SourceDataProvider;
-use cldr_serde::time_zones::bcp47_tzid::Bcp47TzidAliasData;
-use cldr_serde::time_zones::meta_zones::MetazoneAliasData;
-use cldr_serde::time_zones::meta_zones::ZonePeriod;
-use cldr_serde::time_zones::time_zone_names::TimeZoneNames;
 use icu::datetime::provider::time_zones::*;
-use icu::datetime::provider::time_zones::{MetazoneId, TimeZoneBcp47Id};
 use icu::timezone::provider::*;
 use icu_provider::prelude::*;
 use std::collections::BTreeMap;
 use std::collections::HashSet;
+use tinystr::tinystr;
 
 mod convert;
 mod names;
 mod windows;
 
-#[derive(Debug, Copy, Clone)]
-struct CldrTimeZonesData<'a> {
-    pub(crate) time_zone_names_resource: &'a TimeZoneNames,
-    pub(crate) bcp47_tzids_resource: &'a BTreeMap<TimeZoneBcp47Id, Bcp47TzidAliasData>,
-    pub(crate) meta_zone_ids_resource: &'a BTreeMap<MetazoneId, MetazoneAliasData>,
-    pub(crate) meta_zone_periods_resource: &'a BTreeMap<String, ZonePeriod>,
-    pub(crate) tzdb: &'a parse_zoneinfo::table::Table,
-}
-
-macro_rules! impl_data_provider {
-    ($($marker:ident),+) => {
-        $(
-            impl DataProvider<$marker> for SourceDataProvider {
-                fn load(&self, req: DataRequest) -> Result<DataResponse<$marker>, DataError> {
-                    self.check_req::<$marker>(req)?;
-
-                    let resource: &cldr_serde::time_zones::time_zone_names::Resource = self
-                        .cldr()?
-                        .dates("gregorian")
-                        .read_and_parse(req.id.locale, "timeZoneNames.json")?;
-
-                    let time_zone_names_resource = &resource.main.value.dates.time_zone_names;
+// TODO: cache these
+impl SourceDataProvider {
+    /// Returns a map from time zone long identifier to time zone BCP-47 ID.
+    ///
+    /// For example: "America/Chicago" to "uschi"
+    fn compute_bcp47_tzids_btreemap(&self) -> Result<BTreeMap<String, TimeZoneBcp47Id>, DataError> {
```

Comment:
Issue: I really prefer this function working the way it had worked before.

You're now loading `timezone.json` multiple times, sometimes up to 3 times, at some call sites. It also makes this function do more than it is stated to do: it's now "load _and_ compute bcp47 tzids btreemap"

I acknowledge that there are some call sites that load `timezone.json` for only the purpose of invoking this function. I would be okay with a private function on SourceDataProvider that returns the timezone.json resource, which you can then pass down into this function. For example, you can change an existing call site like this:

```rust
// current
let resource: &cldr_serde::time_zones::bcp47_tzid::Resource =
    self.cldr()?.bcp47().read_and_parse("timezone.json")?;
let iana2bcp = &compute_bcp47_tzids_btreemap(&resource.keyword.u.time_zones.values);

// proposed
let resource = self.get_timezone_resource()?;
let iana2bcp = &compute_bcp47_tzids_btreemap(&resource.keyword.u.time_zones.values);
```

Reply:
Abstracting this in a method lets us cache this in the future. I also don't want to repeat the JSON loading everywhere I need the BTreeMap, that's just code duplication.

Reply:
I don't think we should make a separate method until we actually have caching in place, because in the interim it is less efficient because you need to re-load and re-parse the JSON multiple times.

Reply:
It was a separate method before, it's a separate method now. The JSON isn't parsed multiple times, it's cached.